### ISSUE TITLE
Enable RuboCop to enforce a consistent code style and perform static analysis

### ISSUE THREAD

**Issue by Mogztter** [asciidoctor/asciidoctor#2557]
I think it could be a good idea to enforce some guidelines (especially for newcomers and developers with limited knowledge of Ruby).

What do you think ?

**Comment by mojavelinux** [comment]
While I support this generally in Asciidoctor, I'm not ready for this switch for core. I am starting to get use to linters/formatters in general from using them in Antora. However, the source code for Asciidoctor core is formatted in a very particular way and I don't think it's the best use of time atm to change it. I think once we really start messing with the 1.6 code base, that's when we really want to take a serious look at a linter/formatter.

In short, I like the idea, but I don't think the time is right yet.

**Comment by mojavelinux** [comment]
Here's another option: https://github.com/testdouble/standard (standard for Ruby)

**Comment by mojavelinux** [comment]
I'm considering it post 2.0.0. I'll admit, I'm going to be a bit reluctant because I'm so use to the style I'm using, but since it will make it easier on contributors and encourage contribution, I'll at least give serious consideration.

**Comment by mojavelinux** [comment]
I'm starting to play with Rubocop in Jekyll AsciiDoc to get warmed up to it. I decided I don't like standardrb because it forces our hand to much. I think we can groom the ruleset to match close enough to our style that it will be easy to give in on some of the mismatches.

**Comment by mojavelinux** [comment]
I've now been using Rubocop on Asciidoctor PDF for over a year and I can say that I'm now quite comfortable with it and happy with the result. I would like to enable it on Asciidoctor starting with the 2.1 release.

I would like to use the rules defined in the Asciidoctor PDF project as the Asciidoctor Ruby style. The official location for the rules will be Asciidoctor and we'll propagate those out when they get updated. Since adopting the rules requires a lot of unrelated changes in the source code, it's best to wait until the 2.0.x branch is cut before we start working on it.

**Comment by mojavelinux** [comment]
Interesting to note that the [Rubocop](https://rubocop.org/) project uses AsciiDoc and Antora for documentation. See https://docs.rubocop.org/rubocop/1.11/index.html

**Comment by slonopotamus** [comment]
> it's best to wait until the 2.0.x branch is cut before we start working on it.

If you mass-format code *after* forking a branch, you make backporting of fixes from main branch almost impossible to do. So it is possibly better to first mass-format and *then* fork.

**Comment by mojavelinux** [comment]
These changes have been applied, but will not be included in the 2.0.16 release to avoid any delays in repackaging. Rather, they will be applied to the v2.0.x branch once the 2.0.16 release is made.
